doc/source/user/basics.interoperability.rst

```diff
@@ -0,0 +1,427 @@
+
+***************************
+Interoperability with NumPy
+***************************
+
+NumPy's ndarray objects provide both a high-level API for operations on
+array-structured data and a concrete implementation of the API based on
+:ref:`strided in-RAM storage <arrays>`. While this API is powerful and fairly
+general, its concrete implementation has limitations. As datasets grow and NumPy
+becomes used in a variety of new environments and architectures, there are cases
+where the strided in-RAM storage strategy is inappropriate, which has caused
+different libraries to reimplement this API for their own uses. This includes
+GPU arrays (CuPy_), Sparse arrays (`scipy.sparse`, `PyData/Sparse <Sparse_>`_)
+and parallel arrays (Dask_ arrays) as well as various NumPy-like implementations
+in deep learning frameworks, like TensorFlow_ and PyTorch_. Similarly, there are
+many projects that build on top of the NumPy API for labeled and indexed arrays
+(XArray_), automatic differentiation (JAX_), masked arrays (`numpy.ma`),
+physical units (astropy.units_, pint_, unyt_), among others that add additional
+functionality on top of the NumPy API.
+
+Yet, users still want to work with these arrays using the familiar NumPy API and
+re-use existing code with minimal (ideally zero) porting overhead. With this
+goal in mind, various protocols are defined for implementations of
+multi-dimensional arrays with high-level APIs matching NumPy. 
+
+Broadly speaking, there are three groups of features used for interoperability
+with NumPy:
+
+1. Methods of turning a foreign object into an ndarray;
+2. Methods of deferring execution from a NumPy function to another array
+   library;
+3. Methods that use NumPy functions and return an instance of a foreign object.
+
+We describe these features below.
+
+
+1. Using arbitrary objects in NumPy
+-----------------------------------
+
+The first set of interoperability features from the NumPy API allows foreign
+objects to be treated as NumPy arrays whenever possible. When NumPy functions
+encounter a foreign object, they will try (in order):
+
+1. The buffer protocol, described :py:doc:`in the Python C-API documentation
+   <c-api/buffer>`.
+2. The ``__array_interface__`` protocol, described
+   :ref:`in this page <arrays.interface>`. A precursor to Python's buffer
+   protocol, it defines a way to access the contents of a NumPy array from other
+   C extensions.
+3. The ``__array__()`` method, which asks an arbitrary object to convert
+   itself into an array.
```

Comment:
Should we add the `__dlpack__()` method

Reply:
sorry, that is in a different section

Reply:
Do you mean adding a mention of `__dlpack()__` in the `arrays.interface` doc? I can at least add a link to https://data-apis.org/array-api/latest/design_topics/data_interchange.html#dlpack-support

Reply:
Maybe leave it out for now. I thought we had a tracking issue for finishing dlpack, including documentation.

Reply:
So long we don't have first-class consumer API, it doesn't really make sense to expose it any case.  This should be added as part of making that consumer API first-class.